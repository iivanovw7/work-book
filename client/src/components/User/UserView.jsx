import PropTypes from 'prop-types';
import React from 'react';
import dayjs from 'dayjs';
import { opacify } from 'polished';
import styled from 'styled-components';
import Button from '../UI/Button';
import ViewElement from './UserViewElement';
import * as utils from '../../utils';
import * as constants from '../../constants';
/* eslint react/require-default-props: 0 */
/* eslint no-underscore-dangle: 0 */

const StyledWrapper = styled.article`
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
`;
const StyledContainer = styled.div`
	margin: 1em;
	text-align: left;
  width: 100%;
`;

const StyledButtonContainer = styled.div`
	width: 100%;
	display: flex;
	flex-direction: row;
	flex-wrap: no-wrap;
	user-select: none;
	padding: 0.5rem;
	border-radius: 0.2em;
	background-color: ${opacify('0.01', 'rgba(25, 0, 0, 0.1)')};
`;

const UserView = (props) => {
  const {
    history, data, theme, locale, text
  } = props;
  dayjs.locale(locale.slice(0, -1));
  const user = data.getUser;
  const formattedDate = utils.convertUnixTimestamp(
    user.created,
    constants.TIMESTAMP_UNITS.MS,
    '{YYYY} MM-DDTHH:mm:ss SSS [Z] A'
  );

  function handleClick() {
    history.push('/');
  }

  return (
    <StyledWrapper key={user._id}>
      <StyledContainer>
        <h2>User</h2>
        <ViewElement title="Name" value={user.name} />
        <ViewElement title="Surname" value={user.surname} />
        <ViewElement title="Email" value={user.email} />
        <ViewElement title="Phone" value={user.phone} />
        <ViewElement title="Role" value={user.role} />
        <ViewElement
          title="Created"
          value={formattedDate}
        />
      </StyledContainer>
      <StyledButtonContainer>
        <Button
          variant="primary"
          text={text.navigation.back[locale]}
          theme={theme}
          maxWidth="7em"
          handleClick={() => {
            handleClick();
          }}
        />
      </StyledButtonContainer>
    </StyledWrapper>
  );
};

export default UserView;

UserView.propTypes = {
  data: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  locale: PropTypes.string.isRequired,
  text: PropTypes.object.isRequired,
  theme: PropTypes.string.isRequired
};
