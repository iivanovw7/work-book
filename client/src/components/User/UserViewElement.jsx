import PropTypes from 'prop-types';
import React from 'react';

const ViewElement = (props) => {
	const { title, value } = props;

	return (
		<p>
			<strong>
				{title}
				{': '}
			</strong>
			{value}
		</p>
	);
};

export default ViewElement;

ViewElement.propTypes = {
	title: PropTypes.string.isRequired,
	value: PropTypes.string.isRequired
};
