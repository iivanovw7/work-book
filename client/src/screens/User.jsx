import PropTypes from 'prop-types';
import React from 'react';
import useStoreon from 'storeon/react';
import { withRouter } from 'react-router-dom';
import { Col, Grid, Row } from 'react-styled-flexboxgrid';
import styled, { ThemeProvider } from 'styled-components';
import QueryUser from '../queryBlocks/QueryUser';
import TopBar from '../components/TopBar/TopBar';
import { textColor, backgroundColor } from '../theme';
import MobileNavBar from '../components/MobileNavBar/MobileNavbar';
import SideBar from '../components/SideBar/SideBar';
/* eslint react/require-default-props: 0 */
/* eslint no-underscore-dangle: 0 */

const StyledSection = styled.section`
	color: ${textColor};
  background-color: ${backgroundColor};
`;

const StyledSideBar = styled.aside`
	max-width: inherit;
`;

const User = (props) => {
  const { location } = props;
  const { theme, locale, localizedText } = useStoreon('theme', 'locale', 'localizedText');
  const pathName = location.pathname.split('/')[2];

  return (
    <ThemeProvider theme={{ mode: theme }}>
      <StyledSection className="container">
        <section>
          <TopBar locale={locale} text={localizedText} {...props} />
        </section>
        <Grid>
          <Row>
            <Col xs={12} sm={12} md={3} lg={3}>
              <StyledSideBar>
                <SideBar theme={theme} locale={locale} text={localizedText} onlyMobile {...props} />
              </StyledSideBar>
            </Col>
            <Col xs={12} sm={12} md={9} lg={9}>
              <section>
                <QueryUser
                  theme={theme}
                  locale={locale}
                  text={localizedText}
                  id={pathName}
                  {...props}
                />
              </section>
            </Col>
          </Row>
        </Grid>
        <section className="mobile">
          <MobileNavBar theme={theme} locale={locale} text={localizedText} {...props} />
        </section>
      </StyledSection>
    </ThemeProvider>
  );
};

User.propTypes = {
  history: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  userAccess: PropTypes.bool.isRequired
};

export default withRouter(User);
