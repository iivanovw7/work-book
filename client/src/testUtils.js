import { createMemoryHistory } from 'history';
import * as localized from './assets/locales.json';

// Mock browser history
export const history = createMemoryHistory('/dashboard');

// Grid config object for testing
export const mappedSizes = {
	xs: 12,
	sm: 12,
	md: 12,
	lg: 11
};

export const localizedText = {
	locales: localized.locales,
	topBar: localized.topBar,
	login: localized.login,
	navigation: localized.navigation,
	pageNotFound: localized.pageNotFound
};
