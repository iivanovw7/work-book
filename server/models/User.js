import mongoose from 'mongoose';
import bcrypt from 'bcryptjs';

const UserSchema = mongoose.Schema({
  email: { type: String, required: true, unique: true },
  phone: String,
  password: String,
  name: { type: String, required: true },
  surname: { type: String, required: true },
  role: { type: String, required: false, default: 'USER' },
  created: { type: Date, default: Date.now() }
});

UserSchema.methods.comparePassword = function (candidatePassword, cb) {
  bcrypt.compare(candidatePassword, this.password, (err, isMatch) => {
    if (err) return cb(err);
    cb(null, isMatch);
  });
};

export default mongoose.model('User', UserSchema, 'users');
