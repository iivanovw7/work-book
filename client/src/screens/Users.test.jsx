import React from 'react';
import 'cross-fetch/polyfill';
import 'jsdom-global/register';
import { MemoryRouter } from 'react-router-dom';
import * as testUtils from '../testUtils';
import User from './User';

jest.mock('../config/apiURL', () => () => '/api');

jest.mock('storeon/react', () => () => ({
  locale: 'eng',
  theme: 'dark',
  search: ''
}));
jest.mock('../components/Navigation/MobileNavbar', () => () => (
  <div id="MobileNavbar" />
));
jest.mock('../components/Navigation/TopBar', () => () => (
  <div id="TopBar" />
));
jest.mock('../components/Navigation/SideBar', () => () => (
  <div id="SideBar" />
));
jest.mock('../queryBlocks/QueryUser', () => () => (
  <div id="QueryUser" />
));

describe('Testing USER screen: ', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  const props = {
    localizedText: testUtils.localizedText,
    location: {},
    user: {},
    userAccess: true
  };

  const Composition = () => (
    <MemoryRouter>
      <User {...props} />
    </MemoryRouter>
  );

  it('Should render properly with props ', () => {
    const component = testUtils.suppressConsoleWarnings(
      <Composition />,
      'mount',
      console
    );

    expect(component.find('User').length).toEqual(1);
    expect(component.find('#MobileNavbar').length).toEqual(1);
    expect(component.find('#TopBar').length).toEqual(1);
    expect(component.find('#SideBar').length).toEqual(1);
  });

  it('Should render and match snapshot ', () => {
    const component = testUtils.suppressConsoleWarnings(
      <Composition />,
      'mount',
      console
    );

    expect(component).toMatchSnapshot();
  });
});
