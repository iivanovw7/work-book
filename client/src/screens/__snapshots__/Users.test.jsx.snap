// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing [[33mUSER[39m] screen:  Should render and match snapshot  1`] = `ReactWrapper {}`;
